pkg/kubelet/cm/deviceplugin/endpoint_handler.go

```diff
@@ -0,0 +1,202 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deviceplugin
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha"
+)
+
+type endpointHandler interface {
+	NewEndpoint(resourceName, socketPath string) error
+
+	Devices() map[string][]pluginapi.Device
+	Endpoint(resourceName string) (*endpoint, bool)
+
+	Stop() error
+}
+
+type endpointStore interface {
+	Endpoint(resourceName string) (*endpoint, bool)
+	SynchronizedEndpoint(resourceName string) (*synchronizedEndpoint, bool)
+
+	SwapEndpoint(e *endpoint) (*endpoint, bool)
+	DeleteEndpoint(resourceName string) error
+
+	Range(func(string, *synchronizedEndpoint))
+}
+
+type synchronizedEndpoint struct {
+	e    *endpoint
+	done chan string
+}
+
+type endpointStoreImpl struct {
+	sync.Mutex
+	endpoints map[string]*synchronizedEndpoint
+}
+
+type endpointHandlerImpl struct {
+	store    endpointStore
+	callback MonitorCallback
+}
+
+func newEndpointHandlerImpl(f MonitorCallback) *endpointHandlerImpl {
+	return &endpointHandlerImpl{
+		store:    newEndpointStoreImpl(),
+		callback: f,
+	}
+}
+
+func (h *endpointHandlerImpl) NewEndpoint(resourceName, socketPath string) error {
+	var devStore deviceStore = newDeviceStoreImpl(h.callback)
+	if old, ok := h.store.Endpoint(resourceName); ok && old != nil {
+		devStore = old.store()
+	}
+
+	e, err := newEndpoint(socketPath, resourceName, devStore)
+	if err != nil {
+		return err
+	}
+
+	old, ok := h.store.SwapEndpoint(e)
+	go h.trackEnpoint(e)
+
+	if ok && old != nil {
+		old.setStore(newAlwaysEmptyDeviceStore())
+
+		if err = old.stop(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (h *endpointHandlerImpl) trackEnpoint(e *endpoint) {
+	e.run()
+
+	if curr, ok := h.store.SynchronizedEndpoint(e.resourceName); ok && curr.e == e {
+		h.store.DeleteEndpoint(e.resourceName)
+
+		if curr.done != nil {
+			curr.done <- e.resourceName
+		}
+	}
+}
+
+func (h *endpointHandlerImpl) Devices() map[string][]pluginapi.Device {
+	devs := make(map[string][]pluginapi.Device)
+
+	h.store.Range(func(k string, s *synchronizedEndpoint) {
+		devs[k] = s.e.store().Devices()
+	})
+
+	return devs
+}
+
+func (h *endpointHandlerImpl) Endpoint(resourceName string) (*endpoint, bool) {
+	return h.store.Endpoint(resourceName)
+}
+
+func (h *endpointHandlerImpl) Stop() error {
+	syncChan := make(chan string)
+
+	var endpoints []*synchronizedEndpoint
+	h.store.Range(func(k string, s *synchronizedEndpoint) {
+		endpoints = append(endpoints, s)
+	})
+
+	for _, s := range endpoints {
+		s.done = syncChan
+		s.e.stop()
+
+		select {
+		case <-syncChan:
+			continue
+		case <-time.After(time.Second):
+			return fmt.Errorf("Could not stop endpoint %s", s.e.resourceName)
```

Comment:
is it better to continue and return an errorlist after the range finished?

Reply:
I don't think it really matters this error is really here to catch code changes (dependencies, ...) that might affect endpoint stopping.
If you encounter this error it means that the code needs to be changed.